File: ios/story_creator_kit/Sources/story_creator_kit/StoryCreatorKitPlugin.swift
import Flutter
import Foundation
import UIKit

/// Registers the story_creator_kit native API.
///
/// Every method does its heavy work on background queues; Pigeon delivers
/// replies on the main thread.
public class StoryCreatorKitPlugin: NSObject, FlutterPlugin, StoryNativeApi {
  private let events: StoryNativeEvents
  private let jobs = ExportJobRegistry()
  private let workQueue = DispatchQueue(
    label: "story_creator_kit.work", qos: .userInitiated, attributes: .concurrent)

  init(messenger: FlutterBinaryMessenger) {
    events = StoryNativeEvents(binaryMessenger: messenger)
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let messenger = registrar.messenger()
    let instance = StoryCreatorKitPlugin(messenger: messenger)
    StoryNativeApiSetup.setUp(binaryMessenger: messenger, api: instance)
    registrar.publish(instance)
  }

  public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
    jobs.cancelAll()
    StoryNativeApiSetup.setUp(binaryMessenger: registrar.messenger(), api: nil)
  }

  // MARK: - Export

  func exportVideo(request: VideoExportRequest) async throws -> NativeExportResult {
    let job = try jobs.begin(request.jobId)
    defer { jobs.end(request.jobId) }
    let exporter = VideoExporter(request: request, job: job, progress: progressSink(request.jobId))
    return try await exporter.run()
  }

  func exportStillVideo(request: StillVideoExportRequest) async throws -> NativeExportResult {
    let job = try jobs.begin(request.jobId)
    defer { jobs.end(request.jobId) }
    let exporter = StillVideoExporter(
      request: request, job: job, progress: progressSink(request.jobId))
    return try await exporter.run()
  }

  func encodeJpeg(request: JpegEncodeRequest) async throws -> NativeExportResult {
    try await runOffMain { try JpegEncoder.encode(request) }
  }

  func cancelExport(jobId: String) throws {
    jobs.cancel(jobId)
  }

  // MARK: - Inspection

  func probe(path: String) async throws -> NativeMediaProbe {
    try await MediaProber.probe(path: path)
  }

  func waveform(path: String, buckets: Int64) async throws -> [Double] {
    try await WaveformExtractor.extract(path: path, buckets: Int(buckets))
  }

  func thumbnails(
    path: String, timesMs: [Int64], maxWidth: Int64, outputDirectory: String
  ) async throws -> [String] {
    try await ThumbnailGenerator.generate(
      path: path, timesMs: timesMs, maxWidth: Int(maxWidth), outputDirectory: outputDirectory)
  }

  // MARK: - Helpers

  private func runOffMain<T>(_ work: @escaping () throws -> T) async throws -> T {
    try await withCheckedThrowingContinuation { continuation in
      workQueue.async {
        do {
          continuation.resume(returning: try work())
        } catch {
          continuation.resume(throwing: error)
        }
      }
    }
  }

  /// Sends progress to Dart on the main thread, at most every 100 ms or
  /// 1 %, and always for 0 and 1.
  private func progressSink(_ jobId: String) -> (Double) -> Void {
    let throttle = ProgressThrottle()
    return { [weak self] value in
      guard let self, throttle.shouldSend(value) else { return }
      DispatchQueue.main.async {
        Task { @MainActor in
          try? await self.events.onExportProgress(jobId: jobId, progress: value)
        }
      }
    }
  }
}

/// Decides which progress values are worth sending.
final class ProgressThrottle {
  private let lock = NSLock()
  private var lastValue = -1.0
  private var lastTime = Date.distantPast

  func shouldSend(_ value: Double) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    let now = Date()
    let isEdge = value <= 0 || value >= 1
    if isEdge && value == lastValue {
      return false
    }
    if isEdge || value - lastValue >= 0.01 || now.timeIntervalSince(lastTime) >= 0.1 {
      lastValue = value
      lastTime = now
      return true
    }
    return false
  }
}

/// A running export that can be cancelled from another thread.
final class ExportJob {
  let id: String
  private let lock = NSLock()
  private var cancelled = false
  private var onCancel: [() -> Void] = []

  init(id: String) {
    self.id = id
  }

  var isCancelled: Bool {
    lock.lock()
    defer { lock.unlock() }
    return cancelled
  }

  /// Runs [handler] when the job is cancelled (immediately if it already is).
  func whenCancelled(_ handler: @escaping () -> Void) {
    lock.lock()
    if cancelled {
      lock.unlock()
      handler()
      return
    }
    onCancel.append(handler)
    lock.unlock()
  }

  func cancel() {
    lock.lock()
    if cancelled {
      lock.unlock()
      return
    }
    cancelled = true
    let handlers = onCancel
    onCancel.removeAll()
    lock.unlock()
    handlers.forEach { $0() }
  }

  func throwIfCancelled() throws {
    if isCancelled {
      throw storyError(.cancelled, "Export \(id) was cancelled")
    }
  }
}

/// Jobs by id. A cancel that arrives before the job registered is remembered
/// so the job fails immediately when it starts.
final class ExportJobRegistry {
  private let lock = NSLock()
  private var jobs: [String: ExportJob] = [:]
  private var cancelledEarly: Set<String> = []

  func begin(_ id: String) throws -> ExportJob {
    lock.lock()
    defer { lock.unlock() }
    guard !id.isEmpty else {
      throw storyError(.invalidInput, "jobId is empty")
    }
    guard jobs[id] == nil else {
      throw storyError(.invalidInput, "A job with id \(id) is already running")
    }
    let job = ExportJob(id: id)
    if cancelledEarly.remove(id) != nil {
      job.cancel()
    }
    jobs[id] = job
    return job
  }

  func end(_ id: String) {
    lock.lock()
    jobs[id] = nil
    lock.unlock()
  }

  func cancel(_ id: String) {
    lock.lock()
    let job = jobs[id]
    if job == nil {
      // Remember briefly; the Dart side may cancel right after starting.
      cancelledEarly.insert(id)
      if cancelledEarly.count > 32 {
        cancelledEarly.removeFirst()
      }
    }
    lock.unlock()
    job?.cancel()
  }

  func cancelAll() {
    lock.lock()
    let all = Array(jobs.values)
    lock.unlock()
    all.forEach { $0.cancel() }
  }
}

File: android/src/main/kotlin/com/mabrook/story_creator_kit/StoryCreatorKitPlugin.kt
package com.mabrook.story_creator_kit

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/** Registers the story_creator_kit native API. */
class StoryCreatorKitPlugin : FlutterPlugin {
  private var api: StoryNativeApiImpl? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    val impl = StoryNativeApiImpl(binding.applicationContext, StoryNativeEvents(binding.binaryMessenger))
    api = impl
    StoryNativeApi.setUp(binding.binaryMessenger, impl)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    StoryNativeApi.setUp(binding.binaryMessenger, null)
    api?.dispose()
    api = null
  }
}

/**
 * The native API. Pigeon calls the suspend functions on the main dispatcher;
 * file and codec work moves to [Dispatchers.IO], Transformer runs on the main
 * looper.
 */
internal class StoryNativeApiImpl(
  context: Context,
  private val events: StoryNativeEvents,
) : StoryNativeApi {
  private val exporter = MediaExporter(context)
  private val jobs = ExportJobRegistry()
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

  fun dispose() {
    jobs.cancelAll()
    scope.cancel()
  }

  override suspend fun exportVideo(request: VideoExportRequest): NativeExportResult {
    val job = jobs.begin(request.jobId)
    try {
      return exporter.exportVideo(request, job, progressSink(request.jobId))
    } catch (e: Throwable) {
      throw mapError(e, "Video export")
    } finally {
      jobs.end(request.jobId)
    }
  }

  override suspend fun exportStillVideo(request: StillVideoExportRequest): NativeExportResult {
    val job = jobs.begin(request.jobId)
    try {
      return exporter.exportStill(request, job, progressSink(request.jobId))
    } catch (e: Throwable) {
      throw mapError(e, "Still video export")
    } finally {
      jobs.end(request.jobId)
    }
  }

  override suspend fun encodeJpeg(request: JpegEncodeRequest): NativeExportResult = io("JPEG encoding") {
    StillImageTools.encodeJpeg(request)
  }

  override fun cancelExport(jobId: String) {
    jobs.cancel(jobId)
  }

  override suspend fun probe(path: String): NativeMediaProbe = io("Probe") { MediaProber.probe(path) }

  override suspend fun waveform(path: String, buckets: Long): List<Double> = io("Waveform") {
    WaveformExtractor.extract(path, buckets.toInt())
  }

  override suspend fun thumbnails(
    path: String,
    timesMs: List<Long>,
    maxWidth: Long,
    outputDirectory: String,
  ): List<String> = io("Thumbnails") {
    StillImageTools.thumbnails(path, timesMs, maxWidth.toInt(), outputDirectory)
  }

  private suspend fun <T> io(what: String, block: () -> T): T = withContext(Dispatchers.IO) {
    try {
      block()
    } catch (e: Throwable) {
      throw mapError(e, what)
    }
  }

  /** Sends progress on the main thread, at most every 1 % (always 0 and 1). */
  private fun progressSink(jobId: String): (Double) -> Unit {
    var last = -1.0
    return { value ->
      val edge = value <= 0.0 || value >= 1.0
      if ((edge && value != last) || value - last >= 0.01) {
        last = value
        scope.launch {
          try {
            events.onExportProgress(jobId, value)
          } catch (_: Throwable) {
            // The Dart side may have gone away; progress is best effort.
          }
        }
      }
    }
  }
}

/** A running export that can be cancelled from another thread. */
internal class ExportJob(val id: String) {
  private val lock = Any()
  private var cancelled = false
  private val handlers = mutableListOf<() -> Unit>()

  val isCancelled: Boolean
    get() = synchronized(lock) { cancelled }

  fun whenCancelled(handler: () -> Unit) {
    val runNow = synchronized(lock) {
      if (!cancelled) handlers.add(handler)
      cancelled
    }
    if (runNow) handler()
  }

  fun cancel() {
    val toRun = synchronized(lock) {
      if (cancelled) return
      cancelled = true
      handlers.toList().also { handlers.clear() }
    }
    toRun.forEach { it() }
  }

  fun throwIfCancelled() {
    if (isCancelled) throw storyError(ErrorCodes.CANCELLED, "Export $id was cancelled")
  }
}

/** Jobs by id; a cancel that arrives before its job starts is remembered. */
internal class ExportJobRegistry {
  private val jobs = HashMap<String, ExportJob>()
  private val cancelledEarly = LinkedHashSet<String>()

  @Synchronized
  fun begin(id: String): ExportJob {
    if (id.isEmpty()) throw storyError(ErrorCodes.INVALID_INPUT, "jobId is empty")
    if (jobs.containsKey(id)) throw storyError(ErrorCodes.INVALID_INPUT, "A job with id $id is already running")
    val job = ExportJob(id)
    if (cancelledEarly.remove(id)) job.cancel()
    jobs[id] = job
    return job
  }

  @Synchronized
  fun end(id: String) {
    jobs.remove(id)
  }

  fun cancel(id: String) {
    val job = synchronized(this) {
      jobs[id] ?: run {
        cancelledEarly.add(id)
        if (cancelledEarly.size > 32) cancelledEarly.remove(cancelledEarly.first())
        null
      }
    }
    job?.cancel()
  }

  fun cancelAll() {
    val all = synchronized(this) { jobs.values.toList() }
    all.forEach { it.cancel() }
  }
}
